### PR TITLE
[WIP] Generic primary key implementation

### LINKED ISSUE #36: Generic PK field that references real PK field
Currently tortoise can work normally if your PK is int field named `id`. It should instead work with `pk` field like in django, that references real PK field that can have various types and names

### PR DESCRIPTION
**Pull request by abondar**
Closes #36 

Hey @grigi 
I wrote down this implementation and, surprisingly, it's even works (on sqlite).
I'll try to do cleanup on it in following days and write some more tests, but could you review it by then?
This PR carries many changes and I am afraid there could be some places which I could have done better, but missed them in all this work.

I hope you with fresh view will be able to propose some changes for the better here

### CODE REVIEW

**Comment by grigi** [comment]
Wow, this PR does a lot. Adds new datatypes even, and pg-specific optimisations, and makes uuid PK-able :-)
Can you also make CharField PK-able too?

**Comment by abondar** [comment]
Well, it's already should be pk-able, like any hashable value field. It's just that I was already beaten on last line of code and decided to write tests for other field types later

File: tortoise/backends/sqlite/client.py [hunk 1]
```diff
@@ -90,7 +95,10 @@ def _in_transaction(self) -> "TransactionWrapper":
     async def execute_query(self, query: str) -> List[dict]:
         async with self.acquire_connection() as connection:
             self.log.debug(query)
-            return [dict(row) for row in await connection.execute_fetchall(query)]
+            print(query)
+            res = [dict(row) for row in await connection.execute_fetchall(query)]
+            print(res)
```
Comment (grigi) [inline tortoise/backends/sqlite/client.py]: remove prints when you are done.

File: tortoise/fields.py [hunk 2]
```diff
@@ -303,6 +306,28 @@ def to_python_value(
...
+class UUIDField(Field):
+    """
+    UUID Field
+
+    This field can store uuid value. Postgresql will store value with
+    native data type, while others will store it as string
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(type=UUID, *args, **kwargs)
```
Comment (grigi) [inline tortoise/fields.py]: Should generated not also be auto-set when pk=True?

**Comment by grigi** [comment]
@abondar Do you need help in completing this? It has been sitting around for four weeks now?

**Comment by abondar** [comment]
Hi, sorry for dragging this for so long, couldn't find time to finish it and now I am on vacation until middle of may. 
If you have courage and time I would be happy if you will finish it, otherwise I am going to finish it after coming back

**Comment by grigi** [comment]
Cool, enjoy your vacation :smile: 
I'll have a crack at some of it, partly as I want to understand everything here :+1:

**Comment by grigi** [comment]
@abondar I rebased on latest master, fixed the linting/typing related issues, fixed the M2M table creation to use the right FK key type, and fixed postgres to work again.

I see there is a new attribute `fetch_inserted` that seems like it should control wether we fetch the inserted values, but I wonder why this is an attribute, as we should do the most optimal variety of SQL for insert and getting generated value?

I also see you are trying to replace the field classes for postgres, so we can use the native types easier, I didn't try and implement an auto-replace yet.

And, we need quite a few more tests and lots of documentation changes.

**Comment by grigi** [comment]
Could we just finish the test cases for primary keys, and then merge this without the replacing field classes on a per-db basis? This PR does enough and I feel the field refactoring should be a separate issue? Because it is quite large as well.

**Comment by grigi** [comment]
Turns out both MySQL and SQLite handles ROWID primary key tables and non-ROWID tables very differently.
Also the current implementation will force a ROWID on those tables, and then always collect it afterwards.
This is OK, but causes problems for PostreSQL because it can generate other types on the DB, e.g. UUID. What makes it better right now is that we don't support those cases.

This is one of the many things that is going to complicate the field refactoring #97

**Comment by grigi** [comment]
@abondar Please review my changes.

I removed the custom PostreSQL uuid field type, as until we have parametrized queries for everything, we are likely to run into escaping issues when the to_db_value doesn't treat it as a string. PostgreSQL seems smart enough to parse UUIDs so for now it is a working limitation.

Also, have not updated the documentation.

From a code POV I'm happy with what we have here now.

Also, can we discuss what is to happen for v0.12.0 ?

**Comment by abondar** [comment]
Thank you!
Changes looks all good to me, so I think the only concern left is documentation updates. 
If you don't have opportunity to write it, I think I'll try to manage some time on next week for it.

Regarding 0.12.0 I think this PR is big enough to be sole member of it.

**Comment by grigi** [comment]
?? That last commit broke stuff spectacularly in Travis, but I don't know why. as it doesn't happen locally.
In the MySQL test runs it is emitting PG sql, so I suspect some sort of resource leak?

**Comment by grigi** [comment]
Ok, that looks much better. 
Time has been very short on my side, but it appears to be getting a bit better.

**Comment by grigi** [comment]
@abondar Please vet the docs

File: docs/models.rst [hunk 3]
```diff
@@ -65,20 +65,35 @@ Every model should be derived from base model. You also can derive from your own
...
-Sadly, currently only simple integer primary key is supported, there is plans to enhance this in the not too distant future.
+Any of these are valid primary key definitions in a Model:
+
+.. code-block:: python3
+
+    id = fields.IntField(pk=True)
+
+    checksum = fields.CharField(pk=True)
+
+    guid = fields.UUIDField(pk=True)
```
Comment (abondar) [inline docs/models.rst]: I think we can also mention, that filtering by `.filter(pk__in=...)` or whatever other filters is also supported

**Pull request merged**